src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java

```diff
@@ -0,0 +1,636 @@
+                return SENSITIVITY_MEDIUM;
+            } if (ExtendedOptions.SENSITIVITY_LOW.matches(watchEventModifier)) {
+                return SENSITIVITY_LOW;
+            } else {
+                throw new UnsupportedOperationException(watchEventModifier.name());
+            }
+        }
+
+        public static EnumSet<WatchModifier> setOf(final WatchEvent.Modifier[] modifiers) {
+            final EnumSet<WatchModifier> modifierSet = EnumSet.noneOf(WatchModifier.class);
+            for (final WatchEvent.Modifier modifier : modifiers) {
+                if (modifier == null)
+                    throw new NullPointerException("An element in modifier set is 'null'");
+
+                modifierSet.add(WatchModifier.of(modifier));
+            }
+
+            return modifierSet;
+        }
+
+        public static double sensitivityOf(final EnumSet<WatchModifier> modifiers) {
+            if (modifiers.contains(SENSITIVITY_HIGH)) {
+                return 0.1;
+            } else if (modifiers.contains(SENSITIVITY_LOW)) {
+                return 1;
+            } else {
+                return 0.5; // aka SENSITIVITY_MEDIUM
+            }
+        }
+    }
+
+    private static class MacOSXWatchKey extends AbstractWatchKey {
+        private static final Unsafe unsafe = Unsafe.getUnsafe();
+
+        private static final long kFSEventStreamEventFlagRootChanged     = 0x00000020;
+
+        private static final Path relativeRootPath = theFS.getPath("");
+
+        // Full path to this key's watch root directory.
+        private final Path   realRootPath;
+        private final int    realRootPathLength;
+        private final Object rootPathKey;
+
+        // Kinds of events to be reported.
+        private EnumSet<FSEventKind> eventsToWatch;
+
+        // Native FSEventStreamRef as returned by FSEventStreamCreate().
+        private long         eventStreamRef;
+        private final Object eventStreamRefLock = new Object();
+
+        private DirectorySnapshot directorySnapshot;
+
+        MacOSXWatchKey(final MacOSXWatchService watchService,
+                       final UnixPath dir,
+                       final Object rootPathKey) throws IOException {
+            super(dir, watchService);
+            this.realRootPath       = dir.toRealPath().normalize();
+            this.realRootPathLength = realRootPath.toString().length() + 1;
+            this.rootPathKey        = rootPathKey;
+        }
+
+        synchronized void enable(final CFRunLoopThread runLoopThread,
+                                 final EnumSet<FSEventKind> eventsToWatch,
+                                 final EnumSet<WatchModifier> modifierSet) throws IOException {
+            assert(!isValid());
+
+            this.eventsToWatch = eventsToWatch;
+
+            directorySnapshot = DirectorySnapshot.create(getRealRootPath());
+
+            synchronized (eventStreamRefLock) {
+                final int kFSEventStreamCreateFlagWatchRoot  = 0x00000004;
```

Comment:
```suggestion
                final int kFSEventStreamCreateFlagWatchRoot = 0x00000004;
```

Reply:
Thanks! I'll wait for more formatting-related input from @AlanBateman and others and then do all the changes in bulk.